File: tools/ORYAN/packages/bridge/middleware/resonance.py
"""Resonance middleware skeleton for basic header validation."""

import json
from typing import Dict, Optional

REQUIRED_HEADERS = (
    "x-resonance-authority",
    "x-resonance-consent",
    "x-resonance-trace",
)


def validate_resonance_headers(headers: Dict[str, str]) -> Optional[Dict[str, str]]:
    missing = [header for header in REQUIRED_HEADERS if header not in headers]
    if missing:
        return {"error": "missing_resonance_headers", "missing": missing}

    return {
        "authority": headers.get("x-resonance-authority"),
        "consent": headers.get("x-resonance-consent"),
        "trace": headers.get("x-resonance-trace"),
    }


class ResonanceMiddleware:
    """ASGI middleware placeholder that validates Resonance headers."""

    def __init__(self, app):
        self.app = app

    async def __call__(self, scope, receive, send):
        if scope.get("type") != "http":
            await self.app(scope, receive, send)
            return

        raw_headers = scope.get("headers", [])
        headers = {
            key.decode("latin-1"): value.decode("latin-1")
            for key, value in raw_headers
        }

        validation = validate_resonance_headers(headers)
        if validation and "error" in validation:
            payload = {
                "error": "missing_resonance_headers",
                "missing": validation.get("missing", []),
            }
            body = json.dumps(payload).encode("utf-8")
            await send(
                {
                    "type": "http.response.start",
                    "status": 400,
                    "headers": [(b"content-type", b"application/json")],
                }
            )
            await send({"type": "http.response.body", "body": body})
            return

        scope["resonance"] = validation
        await self.app(scope, receive, send)

File: tools/ORYAN/packages/brain/src/middleware/resonance.js
"use strict";

// Resonance middleware skeleton for basic header validation.
function resonanceMiddleware(req, res, next) {
  const required = [
    "x-resonance-authority",
    "x-resonance-consent",
    "x-resonance-trace",
  ];
  const headers = (req && req.headers) || {};
  const missing = required.filter((header) => !headers[header]);

  if (missing.length > 0) {
    if (res && typeof res.status === "function") {
      return res
        .status(400)
        .json({ error: "missing_resonance_headers", missing });
    }

    const err = new Error(
      "Missing Resonance headers: " + missing.join(", ")
    );
    err.statusCode = 400;
    if (typeof next === "function") {
      return next(err);
    }
    return undefined;
  }

  // Attach basic audit metadata for downstream handlers.
  req.resonance = {
    authority: headers["x-resonance-authority"],
    consent: headers["x-resonance-consent"],
    trace: headers["x-resonance-trace"],
  };

  if (typeof next === "function") {
    return next();
  }
  return undefined;
}

module.exports = {
  resonanceMiddleware,
};
